### Natural Language Processing - Homework 1

Submitted by: Apurv Mittal

In [1]:
# Importing necessary libraries
import nltk
from nltk import word_tokenize
from urllib import request
import warnings

In [2]:
#Ignore Warnings

warnings.filterwarnings('ignore')

Reading text files from http://www.gutenberg.org/ebooks/bookshelf/215 based on grade levels.

We will take grade 2 through grade 6 texts for our analysis.

#### Reference to read text files: 
https://www.pythontutorial.net/python-basics/python-read-text-file/

http://www.nltk.org/book/ch03.html


#### Read Grade 2 Book

In [3]:
url = "https://www.gutenberg.org/cache/epub/14668/pg14668.txt"
response = request.urlopen(url)
grade2book = response.read().decode('utf8')


In [4]:
tokens2 = word_tokenize(grade2book)
text2 = nltk.Text(tokens2)

In [5]:
print(text2)

<Text: The Project Gutenberg EBook of McGuffey 's Second...>


#### Read Grade 3 Book

In [6]:
url = "https://www.gutenberg.org/cache/epub/14766/pg14766.txt"
response = request.urlopen(url)
grade3book = response.read().decode('utf8')

In [7]:
tokens3 = word_tokenize(grade3book)
text3 = nltk.Text(tokens3)

In [8]:
print(text3)

<Text: The Project Gutenberg EBook of McGuffey 's Third...>


#### Read Grade 4 Book

In [9]:
url = "https://www.gutenberg.org/cache/epub/14880/pg14880.txt"
response = request.urlopen(url)
grade4book = response.read().decode('utf8')

In [10]:
tokens4 = word_tokenize(grade4book)
text4 = nltk.Text(tokens4)

In [11]:
print(text4)

<Text: The Project Gutenberg EBook of McGuffey 's Fourth...>


#### Read Grade 5 Book

In [12]:
url = "https://www.gutenberg.org/cache/epub/15040/pg15040.txt"
response = request.urlopen(url)
grade5book = response.read().decode('utf8')

In [13]:
tokens5 = word_tokenize(grade5book)
text5 = nltk.Text(tokens5)

In [14]:
print(text5)

<Text: The Project Gutenberg EBook of McGuffey 's Fifth...>


#### Read Grade 6 Book

In [15]:
url = "https://www.gutenberg.org/cache/epub/16751/pg16751.txt"
response = request.urlopen(url)
grade6book = response.read().decode('utf8')

In [16]:
tokens6 = word_tokenize(grade6book)
text6 = nltk.Text(tokens6)

In [17]:
print(text6)

<Text: The Project Gutenberg EBook of McGuffey 's Sixth...>


#### Comparing Lexical Diversity Scores for each grade text

Defining a function to calculate Lexical Diversity with reference from the text book chapter 1.

https://www.nltk.org/book/ch01.html

In [18]:
def lexical_diversity(text):
    return len(set(text))*100 / len(text)

In [19]:
print("Lexical Diversity Score of Grade 2 Book: ",lexical_diversity(text2),"%")
print("Lexical Diversity Score of Grade 3 Book: ",lexical_diversity(text3), "%")
print("Lexical Diversity Score of Grade 4 Book: ",lexical_diversity(text4), "%")
print("Lexical Diversity Score of Grade 5 Book: ",lexical_diversity(text5),"%")
print("Lexical Diversity Score of Grade 6 Book: ",lexical_diversity(text6),"%")

Lexical Diversity Score of Grade 2 Book:  15.88342153068478 %
Lexical Diversity Score of Grade 3 Book:  12.404916695180692 %
Lexical Diversity Score of Grade 4 Book:  12.347441296124474 %
Lexical Diversity Score of Grade 5 Book:  11.28618252181811 %
Lexical Diversity Score of Grade 6 Book:  10.087260442906656 %


The Lexical Diversity score is falling as we go higher in the grades. For example Grade 2 Lexical Diversity is 15.88% while that of Grade 6 book has reduced to 10.08%. This is surprising as one would expect the Lexical Diversity to be higher for higher grade books. The number of unique words should be lot more in the higher grade books, thus leading to higher Lexical Diversity. This needs to be analyzed further as its against our expectations.

#### Vocabulary Size

To further our analysis we have to understand the total vocobulary in each of these books. We will start with comparing the total unique word types (We cannot call them words as it includes punctuations and any other unique characters in the text). Also, we will compare the total length of the book.

In [20]:
print("Total unique word types in Grade 2: ",len(set(text2)), "and Length of Grade 2 Book: ",len(text2))
print("Total unique word types in Grade 3: ",len(set(text3)), "and Length of Grade 3 Book: ",len(text3))
print("Total unique word types in Grade 4: ",len(set(text4)), "and Length of Grade 4 Book: ",len(text4))
print("Total unique word types in Grade 5: ",len(set(text5)), "and Length of Grade 5 Book: ",len(text5))
print("Total unique word types in Grade 6: ",len(set(text6)), "and Length of Grade 6 Book: ",len(text6))

Total unique word types in Grade 2:  4022 and Length of Grade 2 Book:  25322
Total unique word types in Grade 3:  4713 and Length of Grade 3 Book:  37993
Total unique word types in Grade 4:  10380 and Length of Grade 4 Book:  84066
Total unique word types in Grade 5:  14290 and Length of Grade 5 Book:  126615
Total unique word types in Grade 6:  17259 and Length of Grade 6 Book:  171097


Upon comparing the vocabulary, we notice the number of unique word type steadily increases with the grade level. For example: Grade 2 unique word types are `4,022`. While that of Grade 6 book its `17,259`. Which is on expected lines. As the grade level increases the vocabulary is increasing as expected.

From above analysis of Vocabulary and the Lexical Diversity, we notice they don't always agree with each other. Even with larger vocabulary we notice lower Lexical Diversity score. To understand the issue better we need to look at the total number of words or length of the text as well. In the above examples, we notice that with the increase in grades the total length of the book is also increasing. 

The increase in text length is leading to lower Lexical Diversity eventhough the number of unique words have also increased many folds with the grade levels. This tells us that we should compare the Lexical Diversity only with the similar text length, otherwise its likely to give us wrong interpretation.

Lexical Diversity is a good measure only when compared against equal text length. In practical scenario though it may not be possible to be able to trim the texts to same lengths. In such cases, we have to compare the vocabulary size a well to know if the unique words increases as the length increases.

We cannot use only vocabulary or lexical diversity on its own without understanding the text length. Thus to conclude, Lexical Diversity with Vocabulary in context of the length of the text provides the better interpretation.

#### - End of Homework 1 - 